typedargs/typeinfo.py

```diff
@@ -75,24 +76,37 @@ def register_type_source(self, source, name=None):
 
         self._lazy_type_sources.append((source, name))
 
-    def convert_to_type(self, value, typename, **kwargs):
+    def convert_to_type(self, value, type_or_name, **kwargs):
         """
-        Convert value to type 'typename'
+        Convert value to type 'type_or_name'
 
         If the conversion routine takes various kwargs to
         modify the conversion process, \\**kwargs is passed
         through to the underlying conversion function
         """
-        try:
-            if isinstance(value, bytearray):
-                return self.convert_from_binary(value, typename, **kwargs)
 
-            typeobj = self.get_type(typename)
+        if isinstance(type_or_name, str) or self.is_known_type(type_or_name):
+            try:
+                if isinstance(value, bytearray):
+                    return self.convert_from_binary(value, type_or_name, **kwargs)
+
+                typeobj = self.get_type(type_or_name)
+
+                conv = typeobj.convert(value, **kwargs)
+                return conv
+
+            except (ValueError, TypeError) as exc:
+                raise ValidationError("Could not convert value", type=type_or_name, value=value, error_message=str(exc))
+        else:
+            if isinstance(value, type_or_name):
+                return value
+
+            if isinstance(value, str):
+                if not callable(getattr(type_or_name, 'FromString', None)):
+                    raise ValidationError("Converting from string to the given type is not supported.", type=type_or_name, value=value)
 
-            conv = typeobj.convert(value, **kwargs)
-            return conv
-        except (ValueError, TypeError) as exc:
-            raise ValidationError("Could not convert value", type=typename, value=value, error_message=str(exc))
+                conv = type_or_name.FromString(value)
+                return conv
```

Comment:
This seems to miss raising an exception if `value` is not an instance of `type_or_name` and also not a string.  Right now it just drops through all of the conditional and returns an implicit None.

Reply:
Test case would be:

```
@docannotate
def test_function() -> int:
    return 5.2
```

This should raise an exception when called from the command line because the returned type is wrong but I believe from the code will just chomp the return value and turn it into None.

Reply:
Right, missed case. Added an error raising and updated a test - https://github.com/iotile/typedargs/commit/e882b2a85f585c728c38575696114f8d0cd6cd95